src/shims/windows/handle.rs

```diff
@@ -0,0 +1,118 @@
+use rustc_target::abi::HasDataLayout;
+use std::mem::variant_count;
+
+use crate::*;
+
+/// A Windows `HANDLE` that represents a resource instead of being null or a pseudohandle.
+///
+/// This is a seperate type from [`Handle`] to simplify the packing and unpacking code.
+#[derive(Clone, Copy)]
+enum RealHandle {
+    Thread(ThreadId),
+}
+
+impl RealHandle {
+    fn discriminant(self) -> u64 {
+        match self {
+            // can't use zero here because all zero handle is invalid
+            Self::Thread(_) => 1,
+        }
+    }
+
+    fn data(self) -> u64 {
+        match self {
+            Self::Thread(thread) => thread.to_u32() as u64,
+        }
+    }
+
+    fn packed_disc_size() -> u64 {
+        (variant_count::<Self>().log2() + 1) as u64
+    }
+
+    fn to_packed(self, bits: u64) -> u64 {
```

Comment:
What is all that packing/unpacking about? This file needs some high-level comment for what happens and for what purpose.

Also, is this carefully implemented (the packing, the constant -2, things like that) to match something Windows does, or some ad-hoc Miri-specific thing? In the first case, please link to documentation. In the second case, please explain why it is okay for Miri to make up its own thing.